pelican/utils.py

```diff
@@ -622,79 +629,117 @@ def process_translations(content_list, order_by=None):
     `False` or `false` will be used as translations, unless all the items with
     the same slug have that metadata.
 
-    For each content_list item, sets the 'translations' attribute.
-
-    order_by can be a string of an attribute or sorting function. If order_by
-    is defined, content will be ordered by that attribute or sorting function.
-    By default, content is ordered by slug.
+    If there is more than one original item for a given slug and if
+    secondary_attr is given, tries to identify an original item and
+    translations for each value of secondary_attr.
 
-    Different content types can have default order_by attributes defined
-    in settings, e.g. PAGES_ORDER_BY='sort-order', in which case `sort-order`
-    should be a defined metadata attribute in each page.
+    For each content_list item, sets the 'translations' attribute.
     """
     content_list.sort(key=attrgetter('slug'))
     grouped_by_slugs = groupby(content_list, attrgetter('slug'))
     index = []
-    translations = []
-
-    def _warn_source_paths(msg, items, *extra):
-        args = [len(items)]
-        args.extend(extra)
-        args.extend((x.source_path for x in items))
-        logger.warning('{}: {}'.format(msg, '\n%s' * len(items)), *args)
+    warnings = []
```

Comment:
Why are you storing (a.k.a. delaying) warnings?

Reply:
I had to separate generating and raising the warnings because the warnings generated in the function call in line 653 only have to be raised if the subsequent if-clause tests false. At that point, storing them all until the end seemed most elegant. However, I can see why not issuing these warnings as soon as possible is undesirable, so I'll change that.

Reply:
Oh, I see. I missed that. Looking at the code again, it should be possible to do it in one pass, no?

```python
if secondary_attr:
    # groupby lang and secondary_attr
    groupfn = lambda item: (getattr(item, 'lang'), getattr(item, secondary_attr))
else:
    # groupby lang
    groupfn = attrgetter('lang')

original_items = get_original_items(items, groupfn, ...)
```

Reply:
If I understand your suggestion correctly, that would imply a strict approach: translations are only identified if they have the same category as the original article. I was wondering whether there might not be use cases where this isn't true. The rest of `process_translations` seems to be rather tolerant in what it allows, so I tried to stick with that, and only use the secondary attribute if there is ambiguity.

Reply:
Yes, but that's what this fix intends to do, right?

If translations can have different category compared to the original, there are some issues :)

Consider this with `default_lang='en'`
```
article1:
 slug: slug
 lang: en
 category: cat

article2 (translation of article1):
 slug: slug
 lang: fr
 category: translation_of_cat

article3 (completely different one):
 slug: slug
 lang: en
 category: some_other_cat
```

How do you match `article2` as translation of `article1`?

Reply:
My idea was the following: if you only have `article1` and `article2`, then `article1` is identified as the original, and so we accept `article2` as a translation, even though the categories mismatch. If you add `article3`, then both `article1` and `article3` are identified as originals, in which case if you have any translations they should have a matching category, otherwise there's no way to select the right original. In this case, `article2` has a different category and so ends up identified as an original article in its own right, even though it isn't in the default language.

(One could write a plugin to use a different secondary attribute to disambiguate with, or we could even add that as a setting.)

Reply:
I think we need a more generalized solution. 
The problem: there should be an `id`-like attribute that is common among translations. 

So, how about a new setting `TRANSLATION_ID` that's a function which would generate that `id` from a content to aggregate items by. By default, it would be `attrgetter('slug')` (current behavior).

If you want that to be `(slug+category)`? Fine, define the function that does that. Or if you want to add a completely separate metadata for this purpose, no problem, define that in articles and set the config appropriately.

PS: for convenience, we can automatically wrap the strings for that setting with `attrgetter`. So that you don't need to do `TRANSLATION_ID = attrgetter('id')` but only `TRANSLATION_ID = 'id'`.

Reply:
I agree, that seems most sensible. And let users set it to `None` or `[]` if they simply want to skip looking for translations.  I think a collection of attributes suffices for the setting. I'm not sure there is a use case for custom functions, and if there is, users could create an appropriate metadata attribute dynamically, right?

Reply:
Dynamically? With a plugin, yes.

A list of attributes sounds fine. I suggested function for flexibility but you're probably right. It's unlikely you need anything besides attributes.

Reply:
Hm, I was thinking of `PATH_METADATA` and `EXTRA_PATH_METADATA`, but I realise now that these aren't so flexible. May have to add a setting to create attributes from other attributes, but that's for another PR.